relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -119,20 +74,26 @@ pub fn is_trace_fully_sampled(
         return Some(false);
     }
 
-    let sampling_result = get_sampling_result(
-        processing_enabled,
-        None,
-        Some(root_project_state),
-        Some(dsc),
-        None,
-    );
-
-    let sampled = match sampling_result {
-        SamplingResult::Keep => true,
-        SamplingResult::Drop(_) => false,
+    if root_project_config.unsupported() {
+        if processing_enabled {
+            relay_log::error!("found unsupported rules even as processing relay");
+        } else {
+            return Some(true);
+        }
+    }
+
+    let adjustment_rate = match root_project_config.mode {
+        SamplingMode::Total => dsc.sample_rate,
+        _ => None,
     };
 
-    Some(sampled)
+    // TODO(tor): pass correct now timestamp
+    let evaluator = SamplingEvaluator::new(Utc::now()).adjust_client_sample_rate(adjustment_rate);
+
+    let rules = root_project_config.filter_rules(RuleType::Trace);
+
+    let sampling_result: SamplingResult = evaluator.match_rules(dsc.trace_id, dsc, rules).into();
+    Some(sampling_result.should_keep())
```

Comment:
Instead of converting to `SamplingResult`, consider to match the `RuleMatchingState` directly.

```suggestion
    Some(match evaluator.match_rules(dsc.trace_id, dsc, rules) {
        RuleMatchingState::SamplingMatch(m) => m.should_keep(),
        _ => false,
    })
```

Reply:
hmm, im not sure if it's obvious that a lack of match should result in `true` here, case in point: in your example you return `false` whereas it should be `true`. Maybe it's better to use the implementation in `SamplingResult` for that?

Reply:
if we return the bool manually we're essentially re-implementing this behaviour:

```rust
impl From<Evaluation> for SamplingResult {
    fn from(value: Evaluation) -> Self {
        match value {
            Evaluation::Matched(sampling_match) => Self::Match(sampling_match),
            Evaluation::Continue(_) => Self::NoMatch,
        }
    }
}
    /// Returns `true` if the event should be kept.
    pub fn should_keep(&self) -> bool {
        match self {
            SamplingResult::Match(sampling_match) => sampling_match.should_keep(),
            // If no rules matched on an event, we want to keep it.
            SamplingResult::NoMatch => true,
            SamplingResult::Pending => true,
        }
    }
```

Reply:
in the original implementation, it turns into `Keep` if there's no match in `fn determine_from_sampling_match`, which becomes 'true' here:

```rust
    let sampled = match sampling_result {
        SamplingResult::Keep => true,
        SamplingResult::Drop(_) => false,
```

Reply:
You're absolutely right, the `false` was an outright typo in my suggestion :) I was slightly concerned with the additional conversion into a sampling result, however my own typo is a good example of why this sort of deduplication should exist. 

Let's keep it. A suggestion on how to format it, though:

```rust
let evaluation = evaluator.match_rules(dsc.trace_id, dsc, rules);
Some(SamplingResult::from(evaluation).should_keep())
```

Reply:
nice, thanks! that looks a lot nicer than what I wrote